File: _en/Computer/Operating_System/Python_Programming/Download_Util.py
''' Download_Util.py
Author: BSS9395
Update: 2022-11-05T21:46:00+08@China-Shanghai+08
Design: Download Util
'''

import threading
import urllib.request


class Download_Thread(threading.Thread):
    def __init__(self, path, file, start_position, part_size):
        super().__init__()
        self._path = path
        self._file = file
        self._start_position = start_position
        self._part_size = part_size
        self._length = 0

    def run(self):
        request = urllib.request.Request(url=self._path, method="GET")
        request.add_header("Accept", "*/*")
        request.add_header("Charset", "utf-8")
        request.add_header("Connection", "Keep-Alive")
        resource = urllib.request.urlopen(request)
        resource.read(self._start_position)
        self._file.seek(self._start_position, 0)
        while self._length < self._part_size:
            data = resource.read(1024)
            if data is None or len(data) <= 0:
                break
            self._file.write(data)
            self._length += len(data)
        resource.close()
        self._file.close()


class Download_Util(object):
    def __init__(self, path, file, thread_count):
        self._path = path
        self._file = file
        self._thread_count = thread_count
        self._threads = []
        self._file_size = 0

    def _Download(self):
        request = urllib.request.Request(url=self._path, method="GET")
        request.add_header("Accept", "*/*")
        request.add_header("Charset", "utf-8")
        request.add_header("Connection", "Keep-Alive")
        resource = urllib.request.urlopen(request)
        self._file_size = int(resource.headers.get("Content-Length", 0))
        resource.close()

        part_size = self._file_size // self._thread_count + 1
        for i in range(self._thread_count):
            file = open(self._file, "wb")    # create file one by one before running the thread.
            start_position = i * part_size
            thread = Download_Thread(self._path, file, start_position, part_size)
            self._threads.append(thread)
            thread.start()

    def _Get_Complete_Rate(self):
        size_sum = 0
        for i in range(self._thread_count):
            size_sum += self._threads[i]._length
        return float(size_sum / self._file_size)

    def _Show_Progress(self):
        rate = self._Get_Complete_Rate()
        print("%0.2f completed" % rate)
        if rate < 1.0:
            self._timer = threading.Timer(0.1, self._Show_Progress)
            self._timer.start()
        else:
            self._timer.cancel()


_util = Download_Util("https://pngimg.com/uploads/sprite/sprite_PNG98768.png", 'photo.png', 3)
_util._Download()
_util._Show_Progress()

File: _en/Computer/Operating_System/Obsolete/Java_Standard_Library/Java_HTTP/Download_Util.java
/* Download_util.java
Author: BSS9395
Update: 2023-03-08T22:48:00+08@China-Shanghai+08
Design: Download Util
*/
import java.io.InputStream;
import java.io.RandomAccessFile;
import java.net.HttpURLConnection;
import java.net.URL;

public class Download_Util {
    static class Download_Thread extends Thread {
        private String _rename = "";
        private String _path = "";
        private int _pos = 0;
        private int _size = 0;
        public int _length = 0;

        public Download_Thread(String rename, String path, int pos, int size) {
            _rename = rename;
            _path = path;
            _pos = pos;
            _size = size;
        }

        @Override
        public void run() {
            try {
                HttpURLConnection connection = Download_Util._Connect(_path);
                InputStream stream = connection.getInputStream();
                stream.skip(_pos);
                RandomAccessFile file = new RandomAccessFile(_rename, "rw");
                file.seek(_pos);

                byte[] buffer = new byte[1024];
                for(int count = 0; _length < _size && (count = stream.read(buffer)) != -1;) {
                    file.write(buffer, 0, count);
                    _length += count;
                }
                stream.close();
                file.close();
            } catch (Exception exception) {
                exception.printStackTrace();
            }

        }
    }

    static public String _accept =
        "  image/gif, image/jpeg, image/pjpeg, image/pjpeg" +
        ", application/x-shockwave-flash, application/xaml+xml" +
        ", application/vnd.ms-xpsdocument, application/x-ms-xbap" +
        ", application/x-ms-application, application/vnd.ms-excel" +
        ", application/vnd.ms-powerpoint, application/msword" +
        ", */*";
    static public String _language = "zh-CN";
    static public String _charset = "UTF-8";
    static HttpURLConnection _Connect(String path)
        throws Exception {
        URL url = new URL(path);
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setConnectTimeout(1000 * 5);
        connection.setRequestMethod("GET");
        connection.setRequestProperty("Accept", Download_Util._accept);
        connection.setRequestProperty("Accept-Language", Download_Util._language);
        connection.setRequestProperty("Charset", Download_Util._charset);
        connection.setRequestProperty("Connection", "Keep-Alive");
        return connection;
    }

    private String _path = "";
    private String _rename = "";
    private Download_Thread[] _threads;
    private int _file_size = 0;

    public Download_Util(String path, String rename, int threads) {
        _path = path;
        _rename = rename;
        _threads = new Download_Thread[threads];
    }

    public void _Download()
        throws Exception {
        System.out.println(_path);
        HttpURLConnection connection = Download_Util._Connect(_path);
        _file_size = connection.getContentLength();
        connection.disconnect();
        int part_size = _file_size / _threads.length + 1;
        RandomAccessFile file = new RandomAccessFile(_rename, "rw");
        file.setLength(_file_size);
        file.close();
        for(int i = 0; i < _threads.length; i += 1) {
            _threads[i] = new Download_Thread(_rename, _path,i * part_size, part_size);
            _threads[i].start();
        }
    }

    public double _Progress() {
        int total = 0;
        for(int i = 0; i < _threads.length; i += 1) {
            total += _threads[i]._length;
        }
        return ((double)total / (double)_file_size);
    }


    static public void main(String[] argv)
        throws Exception {
        Download_Util util = new Download_Util("http://www.galex.caltech.edu/media/images/glx2012-03r_img01.jpg", "ios.png", 4);
        util._Download();
        new Thread(() -> {
            try {
                while (util._Progress() < 1.0) {
                    System.out.printf("Completed: %f%n", util._Progress());
                    Thread.sleep(100);
                }
            }
            catch (Exception exception) {
                exception.printStackTrace();
            }
        }).start();

        System.out.println("at the end.");
    }
}
